### PR TITLE
feat: add full policy escalation manager

### PR DESCRIPTION
**Pull request by md0x**
Signed-off-by: Pablo Maldonado <pablo@umaproject.org>

**Motivation**

Add the FullPolicyEscalationManager

**Summary**

- FullPolicyEscalationManager
- Tests

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

Fixes #XXXX

### CODE REVIEW

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol [hunk 1]
```diff
@@ -0,0 +1,181 @@
+pragma solidity 0.8.16;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./BaseEscalationManager.sol";
+import "../../interfaces/OptimisticAsserterInterface.sol";
+
+enum AssertionBlockMode {
```
Comment (chrismaree) [inline packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:7]: Can this be placed within the contract code block?

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    None, // No assertions are blocked
+    BlockByAssertingCallerAndAsserter, // Assertion are block by asserting caller and asserter pair
+    BlockByAssertingCaller // Assertions are blocked by asserting caller
+}
+
+/**
+ * @title The FullPolicyEscalationManager allows the owner to configure all Escalation Manager policy parameters and store
+ * arbitration resolutions. Optionally, we can enable assertion blocking via a whitelist of assertingCallers or a
+ * combination of a whitelist of assertingCallers and asserters. On the other hand, it allows us to determine if we want
```
Comment (chrismaree) [inline packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:16]: Consider not using the word "us" but rather "owner". Same applies throughout the comments.

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+     * @param time unix timestamp of the price request.
+     * @param ancillaryData arbitrary data appended to a price request to give the voters more info from the caller.
+     * @return int256 representing the resolved price for the given identifier and timestamp.
+     * @dev This function replicates the interface of the corresponding DVM function to allow the user to use his own
+     * dispute arbitration system when arbitrating via the escalation manager in a DVM-compatible manner. Refer to the
+     * UMA Voting and VotingV2 contracts for further details.
+     */
+    function getPrice(
+        bytes32 identifier,
+        uint256 time,
```
Comment (chrismaree) [inline packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:72]: Nit: can this function go first?

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        bytes memory ancillaryData
+    ) public view override returns (int256) {
+        bytes32 requestId = keccak256(abi.encode(identifier, time, ancillaryData));
```
Comment (chrismaree) [inline packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:75]: Nit: why not change this function to take in `assertionId` and then you don't need to declare any of this local stuff within this function. Will make this function a bit cleaner.

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+ * FullPolicyEscalationManager will return a default policy with all values set to false.
+ */
+contract FullPolicyEscalationManager is BaseEscalationManager, Ownable {
+    struct ArbitrationResolution {
+        bool valueSet; // True if the resolution has been set.
+        bool resolution; // True or false depending on the resolution.
+    }
+
+    event EscalationManagerConfigured(
+        bool blockAssertion,
```
Comment (Reinis-FRP) [inline packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:29]: This is `validateDisputers` instead

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol (hunk 3 above)
Comment (md0x) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:72]: This is the first function defined in the file. Not sure if I understand

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol (hunk 5 above)
Comment (md0x) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:29]: Thanks!

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol (hunk 4 above)
Comment (Reinis-FRP) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:75]: I believe this comment relates to `_checkIfAssertionBlocked` where indeed we can better pass `assertionId` and then get struct from storage, not doing it in `getAssertionPolicy`

Comment (md0x) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:75]: Thanks! Fixed: [here](https://github.com/UMAprotocol/protocol/pull/4293/commits/b0d54e8ef794c0cbe03c93d541e83675e44d3cfd)

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol (hunk 3 above)
Comment (chrismaree) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:72]: Sorry. Did the review on my phone. I mean put the get assertion policy function first. Put the most important function at the top :)

**Review by Reinis-FRP — APPROVED** [review]
great job!

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol (hunk 3 above)
Comment (md0x) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:89]: Oh, ok, makes sense. Fixed!

File: packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+ * dispute arbitrated by the Oracle, and whether to restrict who can register disputes via a whitelistedDisputeCallers list.
+ * @dev If nothing is configured using the setters and configureEscalationManager method upon deployment, the
+ * FullPolicyEscalationManager will return a default policy with all values set to false.
+ */
+contract FullPolicyEscalationManager is BaseEscalationManager, Ownable {
+    enum AssertionBlockMode {
+        None, // No assertions are blocked
+        BlockByAssertingCallerAndAsserter, // Assertion are block by asserting caller and asserter pair
+        BlockByAssertingCaller // Assertions are blocked by asserting caller
+    }
```
[the remark this answers is not in the document]
Comment (Reinis-FRP) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:21]: The reason for not having 2 separate booleans is that its not secure to filter by asserters while asserting callers are not checked. This is due to the fact that `asserter` can be provided as any address in `assertTruth`

[the remark this answers is not in the document]
Comment (mrice32) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:21]: That makes sense! What about two booleans and enforcing the rule by reverting if the owner tries to configure it that way? I just found the code to do the checking a bit hard to parse with the enum setup, I also think enums can be a bit more annoying than bools for users.

[the remark this answers is not in the document]
Comment (Reinis-FRP) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:21]: agree the alternative with booleans would be to enforce this in `configureEscalationManager` so that having `blockByAndAsserter` true requires also `blockByAssertingCaller` being true.

[the remark this answers is not in the document]
Comment (mrice32) [reply packages/core/contracts/optimistic-asserter/implementation/escalation-manager/FullPolicyEscalationManager.sol:21]: Yep, exactly.
